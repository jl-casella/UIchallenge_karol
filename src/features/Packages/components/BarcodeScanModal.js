import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'

import Button from '../../../atomics/Button'
import Modal from 'atomics/Modal'
import { Product } from '../types'
import ProductsTable from './ProductsTable'

const ProductsTableShort = styled(ProductsTable)`
  width: 60%;
  margin: 20px auto 40px;
`

const BarcodeScanModalActionsBar = styled.div`
  display: flex;
  justify-content: flex-end;
  margin-top: 20px;
`

const BarcodeScanModal = ({ products, onPackProduct, onModalClose }) => {
  const onAddProductToPackage = React.useCallback(product => {
    onPackProduct(product)
    onModalClose()
  }, [])

  return (
    <Modal title="Barcode scan ambiguous">
      <p>
        Multiple products matching the barcode scan detected. Select the one to
        add to current package.
      </p>
      <ProductsTableShort
        items={products}
        onProductRowClick={onAddProductToPackage}
      />
      <hr />
      <BarcodeScanModalActionsBar>
        <Button onClick={onModalClose}>Cancel</Button>
      </BarcodeScanModalActionsBar>
    </Modal>
  )
}

BarcodeScanModal.propTypes = {
  onModalClose: PropTypes.func.isRequired,
  products: PropTypes.arrayOf(Product).isRequired,
  onPackProduct: PropTypes.func.isRequired
}

export default BarcodeScanModal
