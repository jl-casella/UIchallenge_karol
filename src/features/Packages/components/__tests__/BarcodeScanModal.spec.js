import { noop } from 'lodash'
import React from 'react'
import { fireEvent, render } from '@testing-library/react'

import BarcodeScanModal from '../BarcodeScanModal'

describe('BarcodeScanModal component test', () => {
  const productsMock = [
    {
      id: 1,
      quantity: 5,
      sku: 'green-ball',
      location: 'a1'
    },
    {
      id: 99,
      quantity: 2,
      sku: 'green-ball',
      location: 'a4'
    }
  ]
  it('Should render correctly', () => {
    const { container } = render(
      <BarcodeScanModal
        onModalClose={noop}
        products={[]}
        onPackProduct={noop}
      />
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('Should call close callback after clicking "Cancel" button', () => {
    const spy = jest.fn()
    const { getByText } = render(
      <BarcodeScanModal onModalClose={spy} products={[]} onPackProduct={noop} />
    )

    const cancelButton = getByText('Cancel')
    expect(spy).toHaveBeenCalledTimes(0)

    fireEvent.click(cancelButton)
    expect(spy).toHaveBeenCalledTimes(1)
  })

  it('Should call onPack callback with valid after clicking on a row', () => {
    const spy = jest.fn()
    const { getByText } = render(
      <BarcodeScanModal
        onModalClose={noop}
        products={productsMock}
        onPackProduct={spy}
      />
    )

    const firstRow = getByText('a1')
    expect(spy).toHaveBeenCalledTimes(0)

    fireEvent.click(firstRow)
    expect(spy).toHaveBeenCalledTimes(1)

    const spyCallArgs = spy.mock.calls[0]
    expect(spyCallArgs.length).toBe(1)
    expect(spyCallArgs[0].id).toBe(1)
    expect(spyCallArgs[0].quantity).toBe(5)
    expect(spyCallArgs[0].sku).toBe('green-ball')
    expect(spyCallArgs[0].location).toBe('a1')
  })

  it('Should call close callback after selecting valid row', () => {
    const spy = jest.fn()
    const { getByText } = render(
      <BarcodeScanModal
        onModalClose={spy}
        products={productsMock}
        onPackProduct={noop}
      />
    )

    const firstRow = getByText('a1')
    expect(spy).toHaveBeenCalledTimes(0)

    fireEvent.click(firstRow)
    expect(spy).toHaveBeenCalledTimes(1)
  })
})
