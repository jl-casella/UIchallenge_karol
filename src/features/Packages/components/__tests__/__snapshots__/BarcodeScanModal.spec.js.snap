// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarcodeScanModal component test Should render correctly 1`] = `
<div
  class="Modal__ModalWrapper-sc-1njgidi-0 HFFLe"
>
  <h2
    class="Modal__ModalTitle-sc-1njgidi-1 eTNGhA"
  >
    Barcode scan ambiguous
  </h2>
  <div
    class="Modal__ModalContent-sc-1njgidi-2 jgPtAS"
  >
    <p>
      Multiple products matching the barcode scan detected. Select the one to add to current package.
    </p>
    <table
      class="Table-nnk0cv-0 ecmwxU BarcodeScanModal__ProductsTableShort-h0vdrt-0 cKiFDQ"
    >
      <thead>
        <tr>
          <th>
            SKU
          </th>
          <th
            class="ProductsTable__LocationHeaderCell-sc-1vgup2v-1 gxVAmz"
          >
            Location
          </th>
          <th
            class="ProductsTable__QuantityHeaderCell-sc-1vgup2v-2 gRyJwF"
          >
            Quantity
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <hr />
    <div
      class="BarcodeScanModal__BarcodeScanModalActionsBar-h0vdrt-1 dfJnHa"
    >
      <button
        class="Button-sc-1bop9qx-0 kLvdUe"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;
